frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+	/// - No id is changed unless required to by the difference between the old threshold list and
+	///   the new.
+	/// - ids whose bags change at all are implicitly rebagged into the appropriate bag in the new
+	///   threshold set.
+	#[allow(dead_code)]
+	pub fn migrate(old_thresholds: &[VoteWeight]) -> u32 {
+		// we can't check all preconditions, but we can check one
+		debug_assert!(
+			crate::ListBags::<T>::iter().all(|(threshold, _)| old_thresholds.contains(&threshold)),
+			"not all `bag_upper` currently in storage are members of `old_thresholds`",
+		);
+
+		let old_set: BTreeSet<_> = old_thresholds.iter().copied().collect();
+		let new_set: BTreeSet<_> = T::BagThresholds::get().iter().copied().collect();
+
+		// accounts that need to be rebagged
+		let mut affected_accounts = BTreeSet::new();
+		// track affected old bags to make sure we only iterate them once
+		let mut affected_old_bags = BTreeSet::new();
+
+		let new_bags = new_set.difference(&old_set).copied();
+		// a new bag means that all accounts previously using the old bag's threshold must now
+		// be rebagged
+		for inserted_bag in new_bags {
+			let affected_bag = {
+				// this recreates `notional_bag_for` logic, but with the old thresholds.
+				let idx = old_thresholds.partition_point(|&threshold| inserted_bag > threshold);
+				old_thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+			};
+			if !affected_old_bags.insert(affected_bag) {
+				// If the previous threshold list was [10, 20], and we insert [3, 5], then there's
+				// no point iterating through bag 10 twice.
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(affected_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		let removed_bags = old_set.difference(&new_set).copied();
+		// a removed bag means that all members of that bag must be rebagged
+		for removed_bag in removed_bags.clone() {
+			if !affected_old_bags.insert(removed_bag) {
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(removed_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		// migrate the voters whose bag has changed
+		let weight_of = T::VoteWeightProvider::vote_weight;
+		Self::remove_many(&affected_accounts);
+		let num_affected = affected_accounts.len() as u32;
+		let _ = Self::insert_many(affected_accounts.into_iter(), weight_of);
+
+		// we couldn't previously remove the old bags because both insertion and removal assume that
+		// it's always safe to add a bag if it's not present. Now that that's sorted, we can get rid
+		// of them.
+		//
+		// it's pretty cheap to iterate this again, because both sets are in-memory and require no
+		// lookups.
+		for removed_bag in removed_bags {
+			debug_assert!(
+				!crate::ListNodes::<T>::iter().any(|(_, node)| node.bag_upper == removed_bag),
+				"no id should be present in a removed bag",
+			);
+			crate::ListBags::<T>::remove(removed_bag);
+		}
+
+		debug_assert!(
```

Comment:
this assertion is more aligned with the old system where `bag_upper` of each `ListNode` was not kept. Now that it is, I think we should also check that.

Reply:
And perhaps, we can just call the top level sanity check here. Note that all the `debug_xxx` macros are just using a feature called `debug_assertions`, so you can do `#[cfg(any(feature. = "std, feature = "debug-assertions"))] fn sanity_check() {...}`.

Reply:
b39e1c05e1f1a8cec49c5f04ca551c00aee0a713 -> adds sanity check for each node

Reply:
597e1865a7f8c509b56977428a60be8a9b2d4df1 we now do List::sanity_check which internally check that each node's bag_upper makes sense